vmware-tanzu/tanzu-framework
# CLI: Add context command

### What this PR does / why we need it

1. Add context types in client config
This is needed to later introduce a context command in the core CLI. Also, add conversion between server and context types for backwards compatibility with older core and older plugins.

2. Add context command
This command is used to create, list, use, and delete Tanzu CLI contexts for various control planes.

3. Hide context command behind a feature flag
The context command is hidden behind the feature flag `features.global.context-target`. This shall be removed once all plugins and successfully tested its working.

### Which issue(s) this PR fixes
Fixes #2200

### Describe testing done for PR
Added unit tests for the config changes. Tested the `tanzu context` command manually with the following steps:
1. Download an older version of the Tanzu CLI (`v0.21.0`)
2. Login with a management cluster `tanzu login`
3. List clusters `tanzu cluster list` to ensure that the TKG kubeconfig is valid
4. Upgrade to the new CLI with context command `make install-cli` and list the contexts `tanzu context list` to ensure the new CLI translates servers to contexts
5. Get management cluster `tanzu management-cluster get` & list clusters `tanzu cluster list` to ensure that existing plugins are still working
6. Cleanup the management cluster server `tanzu config server delete`
7. Repeat steps 2 & 3 to ensure that login command is still working
8. Repeat step 6
9. Create a context for the management cluster `tanzu context create` to ensure that the new command is working
10. List clusters `tanzu cluster list` to ensure that the new CLI translates contexts to servers

### Release note
```release-note
Currently, the Tanzu CLI supports storing authentication information on a per server basis (management cluster). This information is created using the login plugin and can be managed using the config server command in the core CLI. As we add more server types like a workload cluster, Tanzu Mission Control endpoint, etc. we need a consistent mechanism to manage the configuration for every server or rather for every context.

Also, we need to move away from the notion of server to contexts. To clarify, a server is just a remote endpoint that can be accessed via the Tanzu CLI (or its plugins). A context would include a server endpoint and any related client side configurations like default output format, feature flags enabled, etc. This can also be extended in future to support a local environment without any server if needed.

Hence, added a context command as part of the core CLI like below to create and manage contexts in the CLI.

Note: This is currently behind a feature flag. To enable the flag please run tanzu config set features.global.context-target true

Migration notes:
* CLI users should start using the `context` command instead of the `login` command.
* Plugin developers should use the `config.GetContext` methods in the core CLI library instead of `config.GetServer` when looking for server information.

~ tanzu context -h
Configure and manage contexts for the Tanzu CLI

Usage:
  tanzu context [command]

Aliases:
  context, ctx, contexts

Available Commands:
  create      Create a Tanzu CLI context
  delete      Delete a context from the config
  get         Display a context from the config
  list        List contexts
  use         Make a context as the current active context for its type

Flags:
  -h, --help   help for context

Use "tanzu context [command] --help" for more information about a command.
```

### PR Checklist

### Additional information for the reviewer
* The change is split into three commits to separate the config file changes, the actual command addition and the feature flag. Would prefer to keep this history for future.
* `docs/cli-architecture.md` contains user facing changes and `docs/plugin_implementation_guide.md` contains developer related changes.
